kernel/aster-nix/src/fs/utils/page_cache.rs

```diff
@@ -85,16 +85,185 @@ impl Debug for PageCache {
+    }
+
+    pub fn set_ra_page(&mut self, page: usize) {
+        self.ra_pages = if page > Self::MAX_RA_PAGES {
+            Self::MAX_RA_PAGES
+        } else if page < Self::INIT_RA_PAGES {
+            Self::INIT_RA_PAGES
+        } else {
+            page
+        };
+    }
+
+    fn init_ra_range(&mut self, idx: usize, npages: usize) -> bool {
+        // Set initial readahead window.
+        self.ra_start = idx;
+        if self.ra_start >= npages {
+            self.ra_start = Self::INVALID_PAGE_IDX;
+            return false;
+        }
+        // Set initial window size. Make sure no overflow.
+        self.ra_size = Self::INIT_RA_PAGES.min(npages - self.ra_start);
+        // Fully pipelined.
+        self.lookahead_index = self.ra_start;
+        true
+    }
+
+    fn update_ra_range(&mut self, npages: usize) -> bool {
+        // Push forward window.
+        self.ra_start += self.ra_size;
+        if self.ra_start >= npages {
+            self.ra_start = Self::INVALID_PAGE_IDX;
+            return false;
+        }
+        // Calculate new window size.
+        self.ra_size = (self.ra_size * 2).min(self.ra_pages);
+        // Make sure no overflow.
+        self.ra_size = self.ra_size.min(npages - self.ra_start);
+        // Fully pipelined.
+        self.lookahead_index = self.ra_start;
+        true
+    }
+
+    fn is_sequential(&self, idx: usize) -> bool {
+        self.prev_page != Self::INVALID_PAGE_IDX
+            && (idx == self.prev_page || idx == self.prev_page + 1)
+    }
+
+    fn readahead_index(&self) -> usize {
+        self.ra_start + self.ra_size
+    }
+
+    fn get_ra_range(&self) -> Option<Range<usize>> {
+        if self.ra_start == Self::INVALID_PAGE_IDX {
+            return None;
+        }
+        Some(self.ra_start..self.ra_start + self.ra_size)
+    }
+
+    fn is_all_req_completed(&self) -> bool {
+        let nreqs = self.waiter.nreqs();
+        if nreqs == 0 {
+            return false;
+        }
+        for i in 0..nreqs {
+            if self.waiter.status(i) == BioStatus::Submit {
+                return false;
+            }
+        }
+        true
+    }
+
+    pub fn wait_for_prev_readahead(
+        &mut self,
+        pages: &mut MutexGuard<LruCache<usize, Page>>,
+        force: bool,
+    ) -> Result<()> {
+        if force && self.waiter.nreqs() == 0 {
+            return_errno!(Errno::EINVAL)
+        }
+        if force || self.is_all_req_completed() {
+            if matches!(self.waiter.wait(), Some(BioStatus::Complete)) {
+                for idx in self.get_ra_range().unwrap() {
+                    if let Some(page) = pages.get_mut(&idx) {
+                        page.set_state(PageState::UpToDate);
+                    }
+                }
+                self.waiter.clear();
+            } else {
+                return_errno!(Errno::EIO)
+            }
+        }
+
+        Ok(())
+    }
+
+    pub fn should_readahead(&mut self, idx: usize, npages: usize) -> bool {
```

Comment:
`npages` represents the maximal number of pages for readahead, maybe we should consider changing its name for better clarity.

Reply:
Fixed.